core/src/main/scala/kafka/log/remote/RemoteLogManager.scala

```diff
@@ -0,0 +1,467 @@
+
+      // todo-sato may need to do this asychronously as this call should not block partition#stop(delete).
+      if (delete) {
+        try {
+          remoteStorageManager.deleteTopicPartition(tp)
+        } catch {
+          case ex: Exception => error(s"Error occurred while deleting topic partition: $tp", ex)
+        }
+      }
+    }
+  }
+
+  class RLMTask(tp: TopicPartition) extends CancellableRunnable with Logging {
+    this.logIdent = s"[RLMTask partition:$tp ] "
+    @volatile private var leaderEpoch: Int = -1
+    @volatile def isLeader: Boolean = leaderEpoch >= 0
+
+    // The highest offset that is already in the local remote index files
+    // When looking for new remote segments, we will only look for the remote segments that contains larger offsets
+    private var readOffset: Long = rlmIndexer.getOrLoadIndexOffset(tp).getOrElse(-1)
+
+    fetchLog(tp).foreach {log => log.updateRemoteIndexHighestOffset(readOffset)}
+
+    def convertToLeader(leaderEpochVal:Int): Unit = {
+      if (leaderEpochVal < 0) throw new KafkaException(s"leaderEpoch value for topic partition $tp can not be negative")
+      leaderEpoch = leaderEpochVal
+    }
+
+    def convertToFollower(): Unit = {
+      leaderEpoch = -1
+    }
+
+    def copyLogSegmentsToRemote(): Unit = {
+      try {
+        fetchLog(tp).foreach { log => {
+          if (isCancelled()) {
+            info(s"Skipping copying log segments as the current task is cancelled")
+            return
+          }
+
+          // LSO indicates the offset below are ready to be consumed(high-watermark or committed)
+          val lso = log.lastStableOffset
+          if(lso < 0) {
+            warn(s"lastStableOffset for partition $tp is $lso, which should not be negative.")
+          } else if (lso > 0 && readOffset < lso) {
+            // copy segments only till the min of high-watermark or stable-offset
+            // remote storage should contain only committed/acked messages
+            val fetchOffset = lso
+            info(s"Checking for segments to copy, readOffset: $readOffset and fetchOffset: $fetchOffset")
+            val activeSegBaseOffset = log.activeSegment.baseOffset
+            val sortedSegments = log.logSegments(readOffset + 1, fetchOffset).toSeq.sortBy(_.baseOffset)
+            val index:Int = sortedSegments.map(x => x.baseOffset).search(activeSegBaseOffset) match {
+              case Found(x) => x
+              case InsertionPoint(y) => y - 1
+            }
+            if (index <= 0) {
+              info(s"No segments found to be copied for partition $tp with read offset: $readOffset and active " +
+                s"baseoffset: $activeSegBaseOffset")
+            } else {
+              sortedSegments.slice(0, index).foreach { segment =>
+                // store locally here as this may get updated after the below if condition is computed as false.
+                val leaderEpochVal = leaderEpoch
+                if (isCancelled() || !isLeader) {
```

Comment:
`isLeader` check is done for each segment and if it is moved from leader to follower or it is removed from replicas then it stops copying the remaining segments and returns from here. If this instance has become a follower then `copyLogSegmentsToRemote` is not done in subsequent runs because of the below check in `run()` method.

```
 if (isLeader) copyLogSegmentsToRemote()
```

Reply:
I see, so this local check avoid copying segments if a reassignment has happened for the topic-partition while `copyLogSegmentsToRemote` is being executed.
Actually I am trying to understand the role of `leaderEpoch` here. Is it for potential de-dupe?